### ISSUE TITLE
Define selfing_rate and cloning_rate

### ISSUE THREAD

**Issue by jeromekelleher** [popsim-consortium/demes-spec#43]
We don't currently have much of a definition of what the selfing_rate and cloning_rate parameters mean. (Related: #33, #12 )

See also the wider discussion of how much complexity in life history processes we can/should try to capture here: https://github.com/apragsdale/demes-paper/issues/5

**Comment by jeromekelleher** [comment]
cc @MattHartfield

**Comment by molpopgen** [comment]
I think that the biological definitions are pretty straight forward.  Both are probabilities on [0, 1].  Selfing is the same individual used as a parent "twice" to generate the same offspring, with meiosis occurring.  Cloning implies a simple copy of the individual, presumably mitosis, presumably with mutation happening.  The mutation step distinguishes cloning from simple forms of "overlapping generations".

The main constraint, I think, is that Pr(cloning) + Pr(selfing) must sum to something 0 <= x<= 1, such that 1-x is the probability of outcrossing, completing our set of "mating event types"?

**Comment by MattHartfield** [comment]
@molpopgen sounds good. With regards to constraints: we can partition reproduction into sex and asex, with frequencies x and 1-x respectively. Sex can be further partitioned into selfing and outcrossing, with frequencies s and 1-s. So the ratios of outcrossing: selfing: asex are x(1-s): xs: (1-x).

**Comment by MattHartfield** [comment]
Hi,

I've taken a look at the spec (https://popsim-consortium.github.io/demes-spec-docs/main/specification.html) and the current paper draft. It does not look like there's yet any information on the 'cloning_rate' and 'selfing_rate' parameters, so I'm limited in what I can advise for now. That said, I used the above discussion to determine how these values are currently implemented in DEMES.

It seems that the definition of selfing_rate and cloning_rate are set so that selfing_rate = Prob(offspring are produced by selfing) and cloning_rate = Prob(offspring are produced clonally), using the definition of selfing and cloning that @molpopgen provided above. Furthermore, there's constraints that cloning_rate + selfing_rate <= 1, and 0 <= cloning_rate, selfing_rate <= 1 (so the probability of outcrossing is 1-clonal_rate-selfing_rate).

This setup seems OK to me, as long as we're clear as to what each parameter means and how it will be used in the simulation program of choice. In particular, selfing_rate is the average fraction of offspring that are produced via selfing per generation, as opposed to the probability of selfing *given sex occurs*. For a forward-in-time simulation like SLiM or fwdpy11, then their implementation seems straightforward; when offspring are produced, then the probability of the reproductive mode is drawn from a multinomial distribution or similar. Incidentally, I'm not sure if this is how the selfing, clonal values in SLiM are used - does anyone know or would it be worth sending an email to Ben Haller to check?

However, these rates are seldom included in coalescent simulations, so it's less clear how these values can be implemented. Selfing increases coalescent rates in a neutral coalescent if sex is not too rare and per-deme population sizes remain fixed over time, however such an approximation breaks down for very low rates of sex. It is also an open question as to how the coalescent process with different selfing, clonal rates is affected when (i) there is a continuous change in population size over time (e.g., exponential growth), or (ii) under selection, especially selective sweeps. (These are certainly questions I'm interested in investigating further, so let me know if anyone is open to collaborating on these topics!). Overall, my gut instinct is to disable clonal_rate and selfing_rate unless the software explicitly uses them, in case it leads to unexpected modelling behaviour.

Finally, it might be worth stating in the manuscript that we only cover these specific selfing, cloning outcomes and not other types of uniparental reproduction (e.g., automixis).

**Comment by grahamgower** [comment]
Hi @MattHartfield. The [SLiM manual](http://benhaller.com/slim/SLiM_Manual.pdf) discusses selfing and cloning in sections 6.3.1 and 6.3.2 (pg 135 in my copy). Both sections are quite short.

**Comment by MattHartfield** [comment]
Hi @grahamgower. Thanks for the reference, but it's unclear from the manual whether the SLiM selfing rate is similar to what is being proposed for DEMES, or whether it has a different interpretation (e.g. selfing rate conditional on sex occurring). Do you know what the usage is?

**Comment by grahamgower** [comment]
Sorry, no. I haven't spent any time thinking about this stuff at all. But I'm very interested in learning the answer, because I'd like to support the selfing and cloning rates in a [demes->SLiM translator](https://github.com/grahamgower/demes-slim).

**Comment by apragsdale** [comment]
I don't have a strong preference about whether 0 <= p_self + p_clone <= 1, or if we set the p_self as the probability of selfing _given_ sexual reproduction (we just need to be consistent and document it). I use the selfing rate in `moments` integration, but don't consider cloning, so that issue hasn't come up there. I'm not aware of other software implementations using `demes` that are including selfing and cloning, so perhaps we have some freedom here to really nail this down before other implementations start using those selfing and cloning rates. 

I've thought a bit (but not that recently) about distributions of TMRCAs under selfing in the multi-deme case (e.g. https://github.com/apragsdale/demography/blob/master/demography/tmrcas.py, which just implements the single locus case), and have extended those same methods to compute two-locus pairwise TMRCAs under selfing that can handle different selfing rates in multiple demes with migration and size changes. It's a bit rough and far from being finished up/submitted. The two-locus stuff and my notes are in a private repo, but happy to share the link if you're interested.

**Comment by MattHartfield** [comment]
Hi @apragsdale, yes I would like to take a look at your notes if you're happy to share!
@grahamgower thanks for posting on the SLiM page, I'm interested in seeing what the reply is.

**Comment by jeromekelleher** [comment]
Thanks all, great to get into the details here. I'm hoping to have a framework for the full spec in place in the next day or two, so hopefully we can start working on some specific text there soon.

**Comment by molpopgen** [comment]
> With regards to constraints: we can partition reproduction into sex and asex, with frequencies x and 1-x respectively. Sex can be further partitioned into selfing and outcrossing, with frequencies s and 1-s. So the ratios of outcrossing: selfing: asex are x(1-s): xs: (1-x).

A solution like this may be easiest to generalize.  Broadly, we have "sexual"/meoitic reproduction and asexual/mitotic.  Within each class, the probabilities would sum to 1?

**Comment by MattHartfield** [comment]
> > With regards to constraints: we can partition reproduction into sex and asex, with frequencies x and 1-x respectively. Sex can be further partitioned into selfing and outcrossing, with frequencies s and 1-s. So the ratios of outcrossing: selfing: asex are x(1-s): xs: (1-x).
> 
> A solution like this may be easiest to generalize. Broadly, we have "sexual"/meoitic reproduction and asexual/mitotic. Within each class, the probabilities would sum to 1?

Yes, it is assumed that 0 <= x, s <= 1.

**Comment by molpopgen** [comment]
> > > With regards to constraints: we can partition reproduction into sex and asex, with frequencies x and 1-x respectively. Sex can be further partitioned into selfing and outcrossing, with frequencies s and 1-s. So the ratios of outcrossing: selfing: asex are x(1-s): xs: (1-x).
> > 
> > A solution like this may be easiest to generalize. Broadly, we have "sexual"/meoitic reproduction and asexual/mitotic. Within each class, the probabilities would sum to 1?
> 
> Yes, it is assumed that 0 <= x, s <= 1.

And within x and s, the sums of all flavors should be 1, so that we have a fairly simple model.

**Comment by MattHartfield** [comment]
Yes, we can use the above framework to partition different states further (although I'm not sure what you had in mind?). However, one point I was wresting with yesterday was how selfing rates can be best defined for a simulation program. The above calculations assume s = rate of selfing given sex occurs, but the DEMES/SLiM framework simplify define the overall expressed selfing fraction per generation. Although partitioning reproduction into different 'sex'/'asex' classes can help determine how different aspects of the reproductive cycle affects genetic diversity in a theoretical analysis, it may be easiest to lump all selfing-type reproductions into one class that will be the 'selfing_rate', and all clonal reproduction into the 'cloning_rate' for simulation/inference purposes. Does this make sense?

These points also reinforce the idea that the 'selfing_rate' and 'clonal_rate' might need to be re-scaled by the program using the DEMES input, in case it uses a different definition than used here.

**Comment by apragsdale** [comment]
Following up on this - I think the best approach is to have sex/asex proportions sum to 1 and then within sex to allow selfing proportion to range between 0 and 1. This leaves us wiggle room in the future to be able to be more detailed in how sexual reproduction takes place (e.g. prob of sib-mating or other in/outbreeding scenarios).

Would we be happy to pin our spec down on this definition?

**Comment by jeromekelleher** [comment]
SGTM

**Comment by grahamgower** [comment]
This is a change to what we assumed previously, right? So we need to change the parsers to allow `selfing_rate + cloning_rate > 1`? (The values assigned to selfing_rate and cloning_rate variables are indepedent)

**Comment by MattHartfield** [comment]
Sounds good to me - I take it that another condition would be for 0 <= selfing_rate, cloning_rate <= 1?

**Comment by molpopgen** [comment]
SGTM

**Comment by grahamgower** [comment]
> Following up on this - I think the best approach is to have sex/asex proportions sum to 1 and then within sex to allow selfing proportion to range between 0 and 1. This leaves us wiggle room in the future to be able to be more detailed in how sexual reproduction takes place (e.g. prob of sib-mating or other in/outbreeding scenarios).
> 
> Would we be happy to pin our spec down on this definition?

This needs prose in the MDM (and fixing in the reference implementation: see #109).
